### PR TITLE
MXNET-1452: fix compiler warnings

### PR DESCRIPTION
**Pull request by olk**
- set compiler option -Werror on UNIX in order to get aware of compiler
  warnings
- warnings:
  * fix unused variable warning
  * disable strncpy truncation warning
  * disable warning related to memcpy initialization
  * disable out of bounds warning
    - warning disabled because of performance reasons
  * disable integer in boolean context warning
    - this warnings needs some addtional attention

## Description ##
(Brief description on what this PR is about)

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at https://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by olk** [comment]
[2020-05-31T06:51:03.663Z] =====412/413 cpp-header files passed check=====
[2020-05-31T06:51:03.663Z] src/profiler/profiler.h: 1 Errors of 1 Categories map={'whitespace': 1}
[2020-05-31T06:51:03.663Z] =====687/687 cpp-source files passed check=====

I don't see the whitespace error in the code.

File: src/profiler/profiler.h [hunk 1]
```diff
@@ -54,13 +54,33 @@ struct static_string {
...
+    // GCC 8 does not detect the missing handling
+#pragma GCC diagnostic push
+#if __GNUC__ >= 8
+#pragma GCC diagnostic ignored "-Wstringop-truncation"
+#endif
       strncpy(&string_[0] + l, s, string_size - l - 1);
+#pragma GCC diagnostic pop
       string_[string_size - 1] = '\0';
     }
   }
```
Comment (leezu) [inline src/profiler/profiler.h:86]: `[2020-05-31T06:50:54.248Z] src/profiler/profiler.h:87:  "private:" should be preceded by a blank line  [whitespace/blank_line] [3]`

Comment (olk) [reply src/profiler/profiler.h:86]: First: could you tell me a link to the lint tool, please (so I can run it before committing).
Second: The lint error has nothing to do with my commit - seams to be a prior formatting mistake.

Should I add a fix for the correct formating?

Comment (leezu) [reply src/profiler/profiler.h:86]: It's from the CI output. I'm not sure why it didn't fail before

**Comment by leezu** [comment]
@mxnet-bot run ci [unix-cpu, windows-gpu]

**Comment by olk** [comment]
@mxnet-bot run ci [windows-gpu]

**Pull request closed without merging**